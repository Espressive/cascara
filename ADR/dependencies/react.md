# React

Our React version is tied to FE's version, we can bump only after they are doing bumping Ract. For now, all the following ecosystem members will be pinned to their current version:

- @types/react v16.8.0
- @types/react-dom v16.8.0
- react v16.13.1
- react-dom v16.13.1
- react-router v5.2.0
- react-router-dom v5.2.0
- react-scripts v4.0.3
- react-test-renderer v16.13.1
