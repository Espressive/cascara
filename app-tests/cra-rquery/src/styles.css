html,
body,
#login {
  margin: 0.5em;
  padding: 0;
}

main {
  display: grid;
  justify-content: center;
  align-content: center;
  background-color: ghostwhite;
}

aside {
  grid-column: 1;
}

aside nav {
  display: flex;
  flex-direction: column;
}

aside nav a {
  line-height: 1.5em;
}

section {
  grid-column: 2;
}
