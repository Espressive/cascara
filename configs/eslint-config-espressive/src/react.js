// "off" or 0 - turn the rule off
// "warn" or 1 - turn the rule on as a warning (doesn't affect exit code)
// "error" or 2 - turn the rule on as an error (exit code will be 1)

module.exports = {
  rules: {
    'react/boolean-prop-naming': 2,
    'react/default-props-match-prop-types': 2,
    'react/destructuring-assignment': 2,
    'react/jsx-fragments': 2,
    'react/jsx-no-duplicate-props': 2,
    'react/jsx-no-useless-fragment': 2,
    'react/jsx-pascal-case': 2,
    'react/jsx-sort-default-props': 2,
    'react/jsx-sort-props': 2,
    'react/no-children-prop': 2,
    'react/no-danger': 2,
    'react/no-find-dom-node': 2,
    'react/no-multi-comp': 2,
    'react/no-string-refs': 2,
    'react/no-this-in-sfc': 2,
    'react/no-typos': 2,
    'react/no-unused-prop-types': 2,
    'react/no-unused-state': 2,
    'react/prefer-stateless-function': 2,
    'react/self-closing-comp': 2,
    'react/sort-prop-types': 2,
    'react/style-prop-object': 2,
  },
};
