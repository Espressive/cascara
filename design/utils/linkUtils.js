/**
 * Sets a safe rel value based on html link properties.
 * @param {object} props - All props for the component (must include target and rel)
 * @see {@link https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/}
 */
const getSafeLinkRel = props => {
  // Always use the supplied 'rel' if it is defined.
  // Otherwise, check if a target is defined supply a safe rel value.
  return props.rel || props.target ? 'noopener noreferrer' : undefined;
};

export { getSafeLinkRel };
