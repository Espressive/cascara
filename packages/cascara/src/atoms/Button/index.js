export { default, propTypes as buttonPropTypes } from './Button';
