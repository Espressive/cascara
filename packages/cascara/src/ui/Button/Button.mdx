# Button: three MDX files

See if this is a good idea.
