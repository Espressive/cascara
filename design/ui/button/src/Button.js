import React, { forwardRef } from 'react';
import pt from 'prop-types';
import classNames from 'classnames/bind';
import { getSafeLinkRel } from '../../../utils/linkUtils.js';
import styles from './Button.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /** Can render as a different tag or component */
  as: pt.oneOfType([pt.string, pt.node]),
  /** Main content of the button */
  content: pt.string.isRequired,
  /** Makes the button take the width of the parent container */
  fluid: pt.bool,
  /** Sets the color type of the button to follow the theme brand color */
  isBrandColor: pt.bool,
  /** Indicates the outcome of clicking the button ('positive', 'negative') */
  outcome: pt.oneOf(['positive', 'negative']),
};

// Need to make sure all of our FDS components use forwardRef()
const Button = forwardRef(
  (
    {
      as = 'button',
      content = 'Default Content',
      fluid = false,
      isBrandColor = false,
      outcome,
      ...rest
    },
    ref
  ) => {
    const className = cx({
      _: true,
      basic: !outcome,
      fluid: fluid,
      negative: outcome === 'negative',
      positive: outcome === 'positive',
    });

    // Maybe later we should add tooling to help us validate that this is
    // either a valid HTML5 tag, or that it is a component. Technically
    // some of that is being done with proptypes.
    const ComponentType = as;

    return (
      	<ComponentType
        ref={ref}
        {...rest}
        className={className}
        rel={getSafeLinkRel(rest)}
      >
        {content}
      </ComponentType>
    );
  }
);

Button.propTypes = propTypes;

export { Button };
