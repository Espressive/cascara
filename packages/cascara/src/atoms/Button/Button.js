import classNames from 'classnames';
import pt from '@espressive/prop-types';
import React, { forwardRef } from 'react';
import { Button as ReakitButton } from 'reakit';
import { InlineIcon } from '@iconify/react';
import { getSafeLinkRel } from '../../lib/linkUtils';
// import styles from './Button.module.scss';

const cx = classNames;

const propTypes = {
  /** Can render as a different tag or component */
  as: pt.oneOfType([pt.string, pt.elementType]),
  /** Main content of the button */
  content: pt.string,
  /** Makes the button take the width of the parent container */
  fluid: pt.string,
  // An SVG Object, which will override the content
  icon: pt.shape({}),
  /** Sets the color type of the button to follow the theme brand color */
  isBrandColor: pt.bool,
  /** Indicates the outcome of clicking the button ('positive', 'negative') */
  outcome: pt.oneOf(['positive', 'negative']),
  /** A button can have multiple sizes */
  size: pt.oneOf(['small', 'regular', 'large']),
};

// We are going to need a utility to help us dealing with refs. When a composable `as`
// prop is being used with a React component, we need to pass the ref directly to
// Reakit. Otherwise, we need to use React.forwardRef()

const Button = forwardRef(
  (
    {
      as = 'button',
      content = 'Default Content',
      fluid = false,
      icon,
      isBrandColor = false,
      outcome,
      size = 'regular',
      ...rest
    },
    ref
  ) => {
    // Leaving this here for when we decide to run our own styles
    // const internalClassName = cx(rest.className, {
    //   _: true,
    //   basic: !outcome,
    //   fluid: fluid,
    //   large: size === 'large',
    //   negative: outcome === 'negative',
    //   positive: outcome === 'positive',
    //   small: size === 'small',
    // });

    const legacyClassname = cx(rest.className, {
      basic: !outcome,
      fluid: fluid,
      icon: icon,
      large: size === 'large',
      negative: outcome === 'negative',
      positive: outcome === 'positive',
      small: size === 'small',
      'ui button': true,
    });

    // FDS-137: use action name for button name if no content is specified
    const buttonText = content || rest.name;

    return (
      <ReakitButton
        {...rest}
        as={as}
        className={legacyClassname}
        ref={ref}
        rel={getSafeLinkRel(rest)}
      >
        {icon ? <InlineIcon icon={icon} /> : buttonText}
      </ReakitButton>
    );
  }
);
Button.propTypes = propTypes;

export default Button;
