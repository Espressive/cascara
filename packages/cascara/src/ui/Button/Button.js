import classNames from 'classnames/bind';
import pt from 'prop-types';
import React, { forwardRef } from 'react';
import { Button as ReakitButton } from 'reakit';
import { getSafeLinkRel } from '../../shared/linkUtils';
import styles from './Button.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /** Can render as a different tag or component */
  as: pt.oneOfType([pt.string, pt.elementType]),
  /** Main content of the button */
  content: pt.string,
  /** Makes the button take the width of the parent container */
  fluid: pt.bool,
  /** Sets the color type of the button to follow the theme brand color */
  isBrandColor: pt.bool,
  /** Indicates the outcome of clicking the button ('positive', 'negative') */
  outcome: pt.oneOf(['positive', 'negative']),
  /** A button can have multiple sizes */
  size: pt.oneOf(['small', 'regular', 'large']),
};

// We are going to need a utility to help us dealing with refs. When a composable `as`
// prop is being used with a React component, we need to pass the ref directly to
// Reakit. Otherwise, we need to use React.forwardRef()

const Button = forwardRef(
  (
    {
      as = 'button',
      content = 'Default Content',
      fluid = false,
      isBrandColor = false,
      outcome,
      size = 'regular',
      ...rest
    },
    ref
  ) => {
    const internalClassName = cx(rest.className, {
      _: true,
      basic: !outcome,
      fluid: fluid,
      large: size === 'large',
      negative: outcome === 'negative',
      positive: outcome === 'positive',
      small: size === 'small',
    });

    return (
      <ReakitButton
        {...rest}
        as={as}
        className={internalClassName}
        ref={ref}
        rel={getSafeLinkRel(rest)}
      >
        {content}
      </ReakitButton>
    );
  }
);
Button.propTypes = propTypes;

export default Button;
