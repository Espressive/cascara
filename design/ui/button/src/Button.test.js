import React from 'react';
import { render } from '@testing-library/react';
import { Button } from './Button';

test('Renders with defaults', () => {
  const snap = render(<Button />).container;

  expect(snap).toMatchSnapshot();

  // This should always be our first test rendering the bare component.
  // We might need to find a way to make this test not fail for a prop-type
  // error if there are missing required props. Eventually we want tests
  // to fail if there are any console errors.
});
