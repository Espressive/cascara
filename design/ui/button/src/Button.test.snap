// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default renders without any props 1`] = `
<div>
  <button
    class="_ basic"
    data-testid="default"
    type="button"
  >
    test
  </button>
</div>
`;
